torch_xla/csrc/ops/einsum_backward.cpp

```diff
@@ -0,0 +1,152 @@
+      equation.substr(split_index_one + 1, split_index_two);
+  std::string equation_output =
+      equation.substr(split_index_two, equation.size());
+
+  std::string equation_one =
+      equation_input_one + "," + equation_output + "->" + equation_input_two;
+  std::string equation_two =
+      equation_output + "," + equation_input_two + "->" + equation_input_one;
+
+  std::vector<xla::XlaOp> ops;
+  if (inputs.size() == 1) {
+    ops.push_back(xla::Einsum(grad_output, equation));
+  } else if (inputs.size() == 2) {
+    ops.push_back(xla::Einsum(
+        inputs[0], grad_output, equation_one,
+        xla::PrecisionConfig::Precision::PrecisionConfig_Precision_DEFAULT,
+        XlaHelpers::PromoteType(XlaHelpers::TypeOfXlaOp(inputs[0]),
+                                XlaHelpers::TypeOfXlaOp(grad_output))));
+    ops.push_back(xla::Einsum(
+        grad_output, inputs[1], equation_two,
+        xla::PrecisionConfig::Precision::PrecisionConfig_Precision_DEFAULT,
+        XlaHelpers::PromoteType(XlaHelpers::TypeOfXlaOp(grad_output),
+                                XlaHelpers::TypeOfXlaOp(inputs[1]))));
+  }
+
+  return ops;
+}
+
+xla::Shape NodeOutputShapes(const torch::lazy::Value& grad_output,
+                            const torch::lazy::OpList& inputs,
+                            const std::string& equation) {
+  auto lower_for_shape_one_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    return BuildEinsumBackward(
+        operands[0],
+        std::vector<xla::XlaOp>(operands.begin() + 1, operands.end()),
+        equation)[0];
+  };
+
+  auto lower_for_shape_two_fn =
+      [&](absl::Span<const xla::XlaOp> operands) -> xla::XlaOp {
+    return BuildEinsumBackward(
+        operands[0],
+        std::vector<xla::XlaOp>(operands.begin() + 1, operands.end()),
+        equation)[1];
+  };
+
+  std::vector<xla::Shape> input_shapes;
+  input_shapes.push_back(GetXlaShape(grad_output));
+  for (auto const& op : inputs) {
+    input_shapes.push_back(GetXlaShape(op));
+  }
+
+  xla::Shape output_shape;
+  auto shape_one =
+      InferOutputShape(absl::MakeSpan(input_shapes), lower_for_shape_one_fn);
+
+  if (inputs.size() != 1) {
+    auto shape_two =
+        InferOutputShape(absl::MakeSpan(input_shapes), lower_for_shape_two_fn);
+    output_shape = xla::ShapeUtil::MakeTupleShape({shape_one, shape_two});
+  } else {
+    output_shape = shape_one;
+  }
+
+  return output_shape;
+}
+
+torch::lazy::OpList BuildOpList(const torch::lazy::OpList& inputs,
+                                const torch::lazy::Value& grad_output) {
+  std::cout << "steventktest Building OpList" << std::endl;
+  for (const auto& inp : inputs) {
+    std::cout << "steventktest input: " << inp->ToString() << std::endl;
+  }
+  std::cout << "steventktest grad_output: " << grad_output->ToString() << std::endl;
+
+  std::vector<torch::lazy::Value> values;
+  values.reserve(inputs.size() + 1);
+  values.insert(values.begin(), grad_output);
```

Comment:
question, why don't you just use `push_back`? `reserve` just do some early memory preallocation. The size of the values is still 0 I think unless you inject anything to it

Reply:
check https://github.com/pytorch/xla/blob/3d24d955b6121289a3c8bb86eda541fca7a0d69f/torch_xla/csrc/ops/amp_foreach_non_finite_check_and_unscale.cpp#L41 for example, which also takes a OpList as input

Reply:
After following that example, I'm getting past the previous error. Now running into 
```
INVALID_ARGUMENT: Expected one "->" in einsum_config
```
It looks like there's a problem with the logic parsing the string in `BuildEinsumBackward`. I'll add some more print statements to figure out what's going on there.